File: data_structures/queue_stack/valid_parentheses.py
# (),{},[]가 담긴 문자열이 주어지면 괄호가 제대로 닫혔는지 검사하는 문제는 대부분 Stack 구조를 사용하라는 문제라 외워두면 좋다.

# 예제 1
# input: s=")("
# output: False
#
# 예제 2
# input: s="{()[]}"
# output: True
#
# 문제
# s:'{(([]))[]}'
# stack: []
# p: '[]'

def isValid(s):
    stack = [] # python 에서 stack 은 기본 컬랙션 arrayList 로 구현되어 있다
    for p in s:
        if p == '(':
            stack.append(')')
        elif p == '{':
            stack.append('}')
        elif p == '[':
            stack.append(']')
        elif not stack or stack.pop() != p:
            return False
    return not stack

File: javalang/queuestack/ValidParentheses.java
// LeetCode 20 ValidParentheses
package javalang.queuestack;

import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.Map;

public class ValidParentheses {

    public static boolean isValid(String s) {
        Map<Character, Character> map = new HashMap<>();
        map.put(')', '(');
        map.put(']', '[');
        map.put('}', '{');

        ArrayDeque<Character> stack = new ArrayDeque<>();

        for (char c : s.toCharArray()) {
            // 닫는 괄호인 경우
            if (map.containsKey(c)) {
                 // 스택이 비었거나 짝이 안 맞으면 실패
                if (stack.isEmpty() || stack.pop() != map.get(c)) {
                    return false;
                }
            }
            // 여는 괄호인 경우
            else {
                stack.push(c);
            }
        }

        return stack.isEmpty();
    }

    // -------------------------------------
    public static void main(String[] args) {
        String[] tests = {
            "(]",
            "{[()]}"
        };

        for (String test : tests) {
            System.out.println(test + " -> " + isValid(test));
        }
    }
}